Crash under -[WKModernDownloadProgress initWithDownloadTask:]

pvollan opened the pull request:
#### 9bd58ff623c27306e73f3e65724b4ce29ba86e99
<pre>
Crash under -[WKModernDownloadProgress initWithDownloadTask:]
<a href="https://bugs.webkit.org/show_bug.cgi?id=282185">https://bugs.webkit.org/show_bug.cgi?id=282185</a>
<a href="https://rdar.apple.com/138686924">rdar://138686924</a>

Reviewed by Chris Dumez.

Check that URLs passed to -[WKModernDownloadProgress initWithDownloadTask:] are not null, since that is
a requirement.

* Source/WebKit/NetworkProcess/Downloads/cocoa/DownloadCocoa.mm:
(WebKit::Download::publishProgress):

Canonical link: <a href="https://commits.webkit.org/285813@main">https://commits.webkit.org/285813@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/67cae11740fa6381335296e5cfcdb70e83a7fc33

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73858 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53287 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26669 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/78187 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25096 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62420 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1072 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58074 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16434 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76925 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48226 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63543 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38470 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/45033 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23429 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66591 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21380 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79747 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1175 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/600 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66407 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1318 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63556 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65687 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9576 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7752 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11396 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1139 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/3889 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1168 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/1155 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1174 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/NetworkProcess/Downloads/cocoa/DownloadCocoa.mm:
```diff
@@ -148,6 +148,11 @@
     if (enableModernDownloadProgress()) {
         bool isUsingPlaceholder = useDownloadPlaceholder == WebKit::UseDownloadPlaceholder::Yes;
```
Seems this can be done after the early returns.

cdumez commented on Source/WebKit/NetworkProcess/Downloads/cocoa/DownloadCocoa.mm:
```diff
@@ -148,6 +148,16 @@
     if (enableModernDownloadProgress()) {
         bool isUsingPlaceholder = useDownloadPlaceholder == WebKit::UseDownloadPlaceholder::Yes;
 
+        if (!m_downloadTask.get().currentRequest.URL) {
```
Why are we checking m_downloadTask.get().currentRequest.URL, it doesn't seem to be used below.

pvollan replied:
I included this check here, since this URL is referenced inside `-[WKModernDownloadProgress initWithDownloadTask:]`.

Thanks for reviewing!

cdumez commented on Source/WebKit/NetworkProcess/Downloads/cocoa/DownloadCocoa.mm:
```diff
@@ -148,6 +148,16 @@
+            return;
+        }
+
+        if (!url.isValid()) {
```
Also, checking `url.isValid()` is not quite equivalent to checking that `(NSURL *)url` is null I think. I'd suggest converting the URL to a NSURL early and null checking that.

cdumez commented on Source/WebKit/NetworkProcess/Downloads/cocoa/DownloadCocoa.mm:
```diff
@@ -146,9 +146,20 @@
+            return;
+        }
+
+        NSURL *publishURL = (NSURL *)url;
```
`RetainPtr`

pvollan replied:
Ah, yes, good catch, will fix!

pvollan committed:
Crash under -[WKModernDownloadProgress initWithDownloadTask:]
https://bugs.webkit.org/show_bug.cgi?id=282185
rdar://138686924

Reviewed by Chris Dumez.

Check that URLs passed to -[WKModernDownloadProgress initWithDownloadTask:] are not null, since that is
a requirement.

* Source/WebKit/NetworkProcess/Downloads/cocoa/DownloadCocoa.mm:
(WebKit::Download::publishProgress):

Canonical link: https://commits.webkit.org/285813@main
